### PR TITLE
[FLINK-18709][Coordination] Implement PhysicalSlotProvider

### PR DESCRIPTION
**Pull request by azagrebin**
## What is the purpose of the change

`PhysicalSlotProviderImpl` tries to allocate a physical slot from the available idle cached slots in `SlotPool`.
If it is not possible, it requests a new slot from the SlotPool.

## Brief change log

  - Introduce interface `PhysicalSlotProvider`
  - Implement it with the `PhysicalSlotProviderImpl`

## Verifying this change

Unit tests.

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/PhysicalSlotProviderImplTest.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		CompletableFuture.runAsync(() -> slotPool.close(), mainThreadExecutor).join();
+	}
+
+	@Test
+	public void testBulkSlotAllocationFulfilledWithAvailableSlots()
+			throws InterruptedException, java.util.concurrent.ExecutionException, java.util.concurrent.TimeoutException {
+		PhysicalSlotRequest request = createPhysicalSlotRequest();
+		addSlotToSlotPool();
+		CompletableFuture<PhysicalSlotRequest.Result> slotFuture = allocateSlot(request);
+		PhysicalSlotRequest.Result result = slotFuture.get(TIMEOUT.getSize(), TIMEOUT.getUnit());
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/PhysicalSlotProviderImplTest.java:86]: maybe check the result immediately without delay in this case?

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/PhysicalSlotProvider.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * The provider serves physical slot requests.
+ */
+interface PhysicalSlotProvider {
+
+	/**
+	 * Submit a request to allocate a physical slot.
+	 *
+	 * <p>The physical slot can be either allocated from the slots, already available for the job,
```
[the remark this answers is not in the document]
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/PhysicalSlotProvider.java:33]: I will add `which are` for readability:
`the slots, which are already`

**Comment by azagrebin** [comment]
Thanks for the review @zhuzhurk 
I have addressed the comments

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/PhysicalSlotProviderImplTest.java [hunk 3]
```diff
@@ -78,12 +76,11 @@ public void teardown() {
...
 
 	@Test
-	public void testBulkSlotAllocationFulfilledWithAvailableSlots()
-			throws InterruptedException, java.util.concurrent.ExecutionException, java.util.concurrent.TimeoutException {
+	public void testBulkSlotAllocationFulfilledWithAvailableSlots() throws InterruptedException, ExecutionException {
 		PhysicalSlotRequest request = createPhysicalSlotRequest();
 		addSlotToSlotPool();
 		CompletableFuture<PhysicalSlotRequest.Result> slotFuture = allocateSlot(request);
-		PhysicalSlotRequest.Result result = slotFuture.get(TIMEOUT.getSize(), TIMEOUT.getUnit());
+		PhysicalSlotRequest.Result result = slotFuture.get();
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/PhysicalSlotProviderImplTest.java:83]: What I had thought is to use `slotFuture.getNow(null)` and then doing a non-null check for the returned result.
Then we can say that allocation can be fulfilled immediately without delay if there are available slots.

**Pull request closed without merging**